src/storage/src/hummock/event_handler/refiller.rs

```diff
@@ -0,0 +1,246 @@
+// limitations under the License.
+
+use std::collections::{HashSet, VecDeque};
+use std::ops::DerefMut;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::{ready, Context, Poll};
+use std::time::Duration;
+
+use futures::future::{join_all, try_join_all};
+use futures::{Future, FutureExt};
+use itertools::Itertools;
+use risingwave_hummock_sdk::compaction_group::hummock_version_ext::SstDeltaInfo;
+use risingwave_object_store::object::object_metrics::GLOBAL_OBJECT_STORE_METRICS;
+use tokio::sync::Semaphore;
+use tokio::task::JoinHandle;
+
+use crate::hummock::local_version::pinned_version::PinnedVersion;
+use crate::hummock::{HummockResult, SstableStoreRef, TableHolder};
+use crate::monitor::StoreLocalStatistic;
+
+#[derive(Debug)]
+pub struct CacheRefillConfig {
+    pub timeout: Duration,
+    pub data_refill_levels: HashSet<u32>,
+    pub concurrency: usize,
+}
+
+struct Item {
+    handle: JoinHandle<()>,
+    event: CacheRefillerEvent,
+}
+
+/// A cache refiller for hummock data.
+pub struct CacheRefiller {
+    /// order: old => new
+    queue: VecDeque<Item>,
+
+    config: Arc<CacheRefillConfig>,
+
+    concurrency: Arc<Semaphore>,
+
+    sstable_store: SstableStoreRef,
+}
+
+impl CacheRefiller {
+    pub fn new(config: CacheRefillConfig, sstable_store: SstableStoreRef) -> Self {
+        let config = Arc::new(config);
+        let concurrency = Arc::new(Semaphore::new(config.concurrency));
+        Self {
+            queue: VecDeque::new(),
+            config,
+            concurrency,
+            sstable_store,
+        }
+    }
+
+    pub fn start_cache_refill(
+        &mut self,
+        deltas: Vec<SstDeltaInfo>,
+        pinned_version: Arc<PinnedVersion>,
+        new_pinned_version: PinnedVersion,
+    ) {
+        let task = CacheRefillTask {
+            deltas,
+
+            context: Arc::new(CacheRefillContext {
+                config: self.config.clone(),
+                concurrency: self.concurrency.clone(),
+                sstable_store: self.sstable_store.clone(),
+            }),
+        };
+        let event = CacheRefillerEvent {
+            pinned_version,
+            new_pinned_version,
+        };
+        let handle = tokio::spawn(async move { task.run().await });
+        let item = Item { handle, event };
+        self.queue.push_back(item);
+    }
+
+    pub fn last_new_pinned_version(&self) -> Option<&PinnedVersion> {
+        self.queue.back().map(|item| &item.event.new_pinned_version)
+    }
+
+    pub fn next_event(&mut self) -> NextCacheRefillerEvent<'_> {
+        NextCacheRefillerEvent { refiller: self }
+    }
+}
+
+pub struct NextCacheRefillerEvent<'a> {
+    refiller: &'a mut CacheRefiller,
+}
+
+impl<'a> Future for NextCacheRefillerEvent<'a> {
+    type Output = CacheRefillerEvent;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let refiller = &mut self.deref_mut().refiller;
+
+        if let Some(item) = refiller.queue.front_mut() {
```

Comment:
We may try polling more tasks on the head until any of them is not ready. And then we can merge all the ready tasks into one event. We can use the new_pinned_version of the latest task and the pinned_version of the oldest task as the one in the event.

Reply:
> We may try polling more tasks on the head until any of them is not ready.

Agreed.

> And then we can merge all the ready tasks into one event.

I checked codes using `PinnedVersion` when applying the new one. It's true that currently we don't rely on any middle version. But can we assume that is always true?